tests/interlocked_int/interlocked_int.c

```diff
@@ -0,0 +1,999 @@
+    ASSERT_ARE_EQUAL(int64_t, 2, interlocked_or_64(&destination[1], 0));
+    ASSERT_ARE_EQUAL(int64_t, INT64_MAX, comperand_result[0]);
+    ASSERT_ARE_EQUAL(int64_t, INT64_MIN, comperand_result[1]);
+    ASSERT_IS_TRUE(return_val);
+
+    ///cleanup
+    free((void*)destination);
+    free(comperand_result);
+}
+
+
+/*Tests_SRS_INTERLOCKED_43_039: [ interlocked_compare_exchange_128 shall store the initial value of *destination in *comperand_result regardless of the result of the comparison.` ]*/
+/*Tests_SRS_INTERLOCKED_43_063: [ interlocked_compare_exchange_128 shall return false if *comperand_result does not equal the original value of *destination. ]*/
+TEST_FUNCTION(interlocked_compare_exchange_128_does_not_exchange_when_not_equal)
+{
+    ///arrange
+    volatile int64_t* destination = (volatile int64_t*)malloc(2 * sizeof(int64_t));
+    int64_t* comperand_result = (int64_t*)malloc(2 * sizeof(int64_t));
+
+    destination[0] = INT64_MAX;
+    destination[1] = INT64_MIN;
+    comperand_result[0] = INT64_MIN;
+    comperand_result[1] = INT64_MAX;
+
+    int64_t exchange_high = 2;
+    int64_t exchange_low = 3;
+
+    ///act
+    bool return_val = interlocked_compare_exchange_128(destination, exchange_high, exchange_low, comperand_result);
+
+    ///assert
+    ASSERT_ARE_EQUAL(int64_t, INT64_MAX, interlocked_or_64(&destination[0], 0));
+    ASSERT_ARE_EQUAL(int64_t, INT64_MIN, interlocked_or_64(&destination[1], 0));
+    ASSERT_ARE_EQUAL(int64_t, INT64_MAX, comperand_result[0]);
+    ASSERT_ARE_EQUAL(int64_t, INT64_MIN, comperand_result[1]);
+    ASSERT_IS_FALSE(return_val);
+
+    ///cleanup
+    free((void*)destination);
+    free((void*)comperand_result);
+}
+#endif
+
+/*Tests_SRS_INTERLOCKED_43_009: [interlocked_compare_exchange_16 shall compare the 16-bit integers pointed to by destination and comperand. If they are equal, *destination is set to exchange. These operations are performed atomically.]*/
+/*Tests_SRS_INTERLOCKED_43_040: [ interlocked_compare_exchange_16 shall return the initial value of *destination. ]*/
+TEST_FUNCTION(interlocked_compare_exchange_16_equal_case)
+{
+    ///arrange
+    volatile int16_t destination = INT16_MAX;
+    int16_t comperand = INT16_MAX;
+    int16_t exchange = INT16_MIN;
+
+    ///act
+    int16_t return_val = interlocked_compare_exchange_16(&destination, exchange, comperand);
+
+    ///assert
+    ASSERT_ARE_EQUAL(int16_t, INT16_MIN, interlocked_or_16(&destination, 0));
+    ASSERT_ARE_EQUAL(int16_t, INT16_MAX, return_val);
+}
+
+/*Tests_SRS_INTERLOCKED_43_009: [interlocked_compare_exchange_16 shall compare the 16-bit integers pointed to by destination and comperand. If they are equal, *destination is set to exchange. These operations are performed atomically.]*/
+/*Tests_SRS_INTERLOCKED_43_040: [ interlocked_compare_exchange_16 shall return the initial value of *destination. ]*/
+TEST_FUNCTION(interlocked_compare_exchange_16_not_exchanges_when_equal)
```

Comment:
>when_equal [](start = 60, length = 10)

when_not_equal... ?

Reply:
Yeah that was a find-replace artifact. Fixed it.

---
In reply to: [430689511](https://github.com/Azure/azure-c-util/pull/43#discussion_r430689511) [](ancestors = 430689511)